### PR TITLE
MAKE-823: Interval flushing and benchmarks

### PR DESCRIPTION
**Pull request by julianedwards**

### CODE REVIEW

File: basic_sender.go [hunk 1]
```diff
@@ -20,8 +20,10 @@ import (
 	"google.golang.org/grpc/credentials"
 )
 
-// TODO: figure out ideal size for this
-const defaultMaxBufferSize = 4096
+const (
+	defaultMaxBufferSize int = 1e7
```
Comment (tychoish) [inline basic_sender.go]: this can't be the right size.

File: basic_sender.go [hunk 2]
```diff
@@ -347,6 +365,27 @@ func (b *buildlogger) createNewLog(ts time.Time) error {
 	return nil
 }
 
+func (b *buildlogger) timedFlush() {
+	b.timer = time.NewTimer(b.opts.FlushInterval)
+	defer func() {
+		_ = b.timer.Stop()
```
Comment (tychoish) [inline basic_sender.go]: I feel like we call timer.Stup all the time without doing _ =

File: cmd/benchmarks/run-benchmarks.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	"log"
+
+	"github.com/evergreen-ci/timber"
+)
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	fmt.Println("running basic sender benchmark...")
```
Comment (tychoish) [inline cmd/benchmarks/run-benchmarks.go]: shouldn't this do poplar

File: cmd/benchmarks/run-benchmarks.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+	if err := timber.RunBasicSenderBenchmark(ctx); err != nil {
```
Comment (tychoish) [inline cmd/benchmarks/run-benchmarks.go]: does it make sense to have the benchmark functions public in the timber package.

File: cmd/benchmarks/run-benchmarks.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+		log.Println(err)
```
Comment (tychoish) [inline cmd/benchmarks/run-benchmarks.go]: grip?

File: benchmark.go [hunk 6]
```diff
@@ -0,0 +1,435 @@
+package timber
```
Comment (tychoish) [inline benchmark.go]: I wouldn't put this in the public api of the timber package. maybe make a benchmark subpackage.

File: benchmark.go [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+// sender implementation.
+func RunBasicSenderBenchmark(ctx context.Context) error {
+	srvCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	if err := setupBenchmark(srvCtx); err != nil {
+		return err
+	}
+	defer func() {
+		if err := teardownBenchmark(srvCtx); err != nil {
+			log.Println(err)
```
Comment (tychoish) [inline benchmark.go]: log with grip?

File: benchmark.go [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+		}
+		r.End(time.Since(startAt))
+		r.IncOps(1)
+
+		return nil
+	}
+
+}
+
+func newUUID() string {
```
Comment (tychoish) [inline benchmark.go]: there's a UUID package that we use sometimes for things like this.

File: benchmark.go [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+	b := make([]byte, 16)
+	_, _ = rand.Read(b)
+	return hex.EncodeToString(b)
+}
+
+var seededRand *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+func newRandCharSetString(length int) string {
```
Comment (tychoish) [inline benchmark.go]: should we move this to a util package

File: basic_sender.go (hunk 1 above)
Comment (julianedwards) [reply basic_sender.go]: I thought we agreed to 10 MB?

File: cmd/benchmarks/run-benchmarks.go (hunk 4 above)
Comment (julianedwards) [reply cmd/benchmarks/run-benchmarks.go]: I think so, the benchmarks I wrote require access to private functions in the timber package

File: cmd/benchmarks/run-benchmarks.go (hunk 3 above)
Comment (julianedwards) [reply cmd/benchmarks/run-benchmarks.go]: it does

File: benchmark.go (hunk 9 above)
Comment (julianedwards) [reply benchmark.go]: we can, thought I do not know if it is worth it? it is only two small helper functions

**Pull request merged**